src/spriteloader/makeindexed.cpp

```diff
@@ -0,0 +1,183 @@
+
+#include "../stdafx.h"
+#include "../core/bitmath_func.hpp"
+#include "../core/math_func.hpp"
+#include "../gfx_func.h"
+#include "../table/sprites.h"
+#include "makeindexed.hpp"
+
+#include "../safeguards.h"
+
+/*
+ * PALETTE_BITS reduces the bits-per-channel of 32bpp graphics data to allow faster palette lookups from
+ * a smaller lookup table.
+ *
+ * 6 bpc is chosen as this results in a palette lookup table of 256KiB with adequate fidelty.
+ * In constract, a 5 bpc lookup table would be 32KiB, and 7 bpc would be 2MiB.
+ *
+ * Values in the table are filled as they are first encountered -- larger lookup table means more colour
+ * distance calculations, and is therefore slower.
+ */
+const uint PALETTE_BITS = 6;
+const uint PALETTE_SHIFT = 8 - PALETTE_BITS;
+const uint PALETTE_BITS_MASK = ((1U << PALETTE_BITS) - 1) << PALETTE_SHIFT;
+const uint PALETTE_BITS_OR = (1U << (PALETTE_SHIFT - 1));
+
+using PaletteLookup = std::array<uint8_t, (1U << PALETTE_BITS * 3)>;
+
+/* Palette lookup data, allocated the first time we need to convert from 32bpp to 8bpp. */
+static std::unique_ptr<PaletteLookup> _palette_lookup;
+
+/**
+ * Calculate distance between two colours.
+ * @param col1 First colour.
+ * @param col2 Second colour.
+ * @returns Euclidean distance between col1 and col2.
+ */
+static uint CalculateColourDistance(const Colour &col1, const Colour &col2)
+{
+	/* Euclidean colour distance for sRGB based on https://en.wikipedia.org/wiki/Color_difference#sRGB */
+	int r = (int)col1.r - (int)col2.r;
+	int g = (int)col1.g - (int)col2.g;
+	int b = (int)col1.b - (int)col2.b;
+
+	return (col1.r + col2.r < 256)
+		? ((2 * r * r) + (4 * g * g) + (3 * b * b))
+		: ((3 * r * r) + (4 * g * g) + (2 * b * b));
+}
+
+/**
+ * Find nearest colour palette index for a 32bpp pixel.
+ * @param pixel Pixel to find.
+ * @param remap Whether palette index should fall in the company-colour remap range.
+ * @returns palette index of nearest colour.
+ */
+static uint8_t FindNearestColourIndex(const SpriteLoader::CommonPixel &pixel, bool remap)
+{
+	uint32_t key = (pixel.r >> PALETTE_SHIFT) | (pixel.g >> PALETTE_SHIFT) << PALETTE_BITS | (pixel.b >> PALETTE_SHIFT) << (PALETTE_BITS * 2);
+	if ((*_palette_lookup)[key] != 0) return (*_palette_lookup)[key];
+
+	/* Reduce bits per channel to PALETTE_BITS, and place value in the middle of the reduced range.
+	 * This is to conteract the information lost between bright and dark pixels, e.g if PALETTE_BITS was 2:
+	 *    0 -  63 ->  32
+	 *   64 - 127 ->  96
+	 *  128 - 191 -> 160
+	 *  192 - 255 -> 224
+	 */
+	const Colour col((pixel.r & PALETTE_BITS_MASK) | PALETTE_BITS_OR,(pixel.g & PALETTE_BITS_MASK) | PALETTE_BITS_OR, (pixel.b & PALETTE_BITS_MASK) | PALETTE_BITS_OR);
+
+	uint best_index = 0;
+	uint best_distance = UINT32_MAX;
+
+	if (remap && IsInsideMM(pixel.m, 0xC6, 0xCE)) {
+		for (uint i = 0xC6; i < 0xCE; i++) {
```

Comment:
These numbers are a bit magic. Aren't they used in this context in other places throughout the code?
Similarly what does the 214 below denote (I reckon the start of another recolour range, right?)